File: internal/installers/docker/ubuntu.go
package docker

import (
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"time"

	"github.com/keyxmare/DevBootstrap/internal/installers"
	"github.com/keyxmare/DevBootstrap/internal/runner"
)

// ubuntuInstaller handles Docker installation on Ubuntu/Debian.
type ubuntuInstaller struct {
	*installers.BaseInstaller
}

// newUbuntuInstaller creates a new Ubuntu/Debian Docker installer.
func newUbuntuInstaller(base *installers.BaseInstaller) *ubuntuInstaller {
	return &ubuntuInstaller{BaseInstaller: base}
}

// CheckExisting checks if Docker is already installed.
func (u *ubuntuInstaller) CheckExisting() (installers.AppStatus, string) {
	if u.Runner.CommandExists("docker") {
		version := u.Runner.GetCommandVersion("docker", "--version")
		return installers.StatusInstalled, version
	}
	return installers.StatusNotInstalled, ""
}

// Install installs Docker Engine on Ubuntu/Debian.
func (u *ubuntuInstaller) Install(opts *installers.InstallOptions) *installers.InstallResult {
	result := &installers.InstallResult{Success: true}

	// Step 1: Remove old packages
	u.CLI.PrintSection("Suppression des anciens paquets Docker")
	oldPackages := []string{
		"docker.io", "docker-doc", "docker-compose", "docker-compose-v2",
		"podman-docker", "containerd", "runc",
	}
	u.Runner.Run(
		append([]string{"apt-get", "remove", "-y"}, oldPackages...),
		runner.WithSudo(),
		runner.WithDescription("Suppression des anciens paquets"),
	)

	// Step 2: Install prerequisites
	u.CLI.PrintSection("Installation des prerequis")
	prereqs := []string{"ca-certificates", "curl", "gnupg"}
	prereqResult := u.Runner.Run(
		append([]string{"apt-get", "install", "-y"}, prereqs...),
		runner.WithSudo(),
		runner.WithDescription("Installation des prerequis"),
	)
	if !prereqResult.Success {
		return installers.NewFailureResult("Echec de l'installation des prerequis", prereqResult.Stderr)
	}
	u.CLI.PrintSuccess("Prerequis installes")

	// Step 3: Add Docker's official GPG key
	u.CLI.PrintSection("Configuration du depot Docker")
	u.CLI.PrintInfo("Ajout de la cle GPG Docker...")

	// Create keyrings directory
	u.Runner.Run([]string{"mkdir", "-p", "/etc/apt/keyrings"}, runner.WithSudo())

	// Download and add GPG key
	gpgKeyURL := "https://download.docker.com/linux/ubuntu/gpg"
	gpgKeyPath := "/etc/apt/keyrings/docker.asc"

	// Download GPG key
	tmpGPG := filepath.Join(os.TempDir(), "docker.gpg")
	if !u.Runner.DownloadFile(gpgKeyURL, tmpGPG, runner.WithDescription("Telechargement de la cle GPG")) {
		return installers.NewFailureResult("Echec du telechargement de la cle GPG")
	}

	// Move to keyrings
	u.Runner.Run([]string{"mv", tmpGPG, gpgKeyPath}, runner.WithSudo())
	u.Runner.Run([]string{"chmod", "a+r", gpgKeyPath}, runner.WithSudo())
	u.CLI.PrintSuccess("Cle GPG ajoutee")

	// Step 4: Set up repository
	u.CLI.PrintInfo("Configuration du depot APT...")

	// Get distribution codename
	codename := u.getDistroCodename()
	arch := u.getArch()

	repoLine := fmt.Sprintf("deb [arch=%s signed-by=%s] https://download.docker.com/linux/ubuntu %s stable",
		arch, gpgKeyPath, codename)

	// Write repository file
	repoFile := "/etc/apt/sources.list.d/docker.list"
	u.Runner.Run(
		[]string{"bash", "-c", fmt.Sprintf("echo '%s' > %s", repoLine, repoFile)},
		runner.WithSudo(),
	)
	u.CLI.PrintSuccess("Depot configure")

	// Step 5: Update package index
	u.CLI.PrintSection("Mise a jour des paquets")
	updateResult := u.Runner.Run(
		[]string{"apt-get", "update"},
		runner.WithSudo(),
		runner.WithDescription("Mise a jour de l'index des paquets"),
	)
	if !updateResult.Success {
		result.AddWarning("Echec de la mise a jour de l'index")
	}

	// Step 6: Install Docker packages
	u.CLI.PrintSection("Installation de Docker")
	dockerPackages := []string{
		"docker-ce",
		"docker-ce-cli",
		"containerd.io",
		"docker-buildx-plugin",
	}

	if opts.InstallCompose {
		dockerPackages = append(dockerPackages, "docker-compose-plugin")
	}

	installResult := u.Runner.Run(
		append([]string{"apt-get", "install", "-y"}, dockerPackages...),
		runner.WithSudo(),
		runner.WithDescription("Installation de Docker Engine"),
		runner.WithTimeout(10*time.Minute),
	)

	if !installResult.Success {
		return installers.NewFailureResult("Echec de l'installation de Docker", installResult.Stderr)
	}
	u.CLI.PrintSuccess("Docker installe")

	// Step 7: Add user to docker group
	if opts.AddUserToDockerGroup {
		u.CLI.PrintSection("Configuration du groupe docker")
		username := os.Getenv("USER")
		if username == "" {
			username = os.Getenv("LOGNAME")
		}

		if username != "" && username != "root" {
			u.Runner.Run([]string{"usermod", "-aG", "docker", username}, runner.WithSudo())
			u.CLI.PrintSuccess(fmt.Sprintf("Utilisateur '%s' ajoute au groupe docker", username))
			result.AddWarning("Deconnectez-vous et reconnectez-vous pour utiliser Docker sans sudo")
		}
	}

	// Step 8: Enable and start Docker service
	if opts.StartOnBoot {
		u.CLI.PrintSection("Configuration du service")
		u.Runner.Run([]string{"systemctl", "enable", "docker"}, runner.WithSudo())
		u.Runner.Run([]string{"systemctl", "start", "docker"}, runner.WithSudo())
		u.CLI.PrintSuccess("Service Docker demarre et active au boot")
	}

	// Verify installation
	if u.Verify() {
		version := u.Runner.GetCommandVersion("docker", "--version")
		result.Message = "Docker installe avec succes"
		result.Version = version
	} else {
		result.AddWarning("L'installation semble terminee mais la verification a echoue")
	}

	return result
}

// getDistroCodename returns the distribution codename.
func (u *ubuntuInstaller) getDistroCodename() string {
	result := u.Runner.Run([]string{"lsb_release", "-cs"})
	if result.Success {
		return strings.TrimSpace(result.Stdout)
	}

	// Fallback: read from /etc/os-release
	result = u.Runner.Run([]string{"bash", "-c", ". /etc/os-release && echo $VERSION_CODENAME"})
	if result.Success {
		return strings.TrimSpace(result.Stdout)
	}

	return "jammy" // Default to Ubuntu 22.04
}

// getArch returns the architecture string for APT.
func (u *ubuntuInstaller) getArch() string {
	result := u.Runner.Run([]string{"dpkg", "--print-architecture"})
	if result.Success {
		return strings.TrimSpace(result.Stdout)
	}
	return "amd64"
}

// Verify verifies the Docker installation.
func (u *ubuntuInstaller) Verify() bool {
	result := u.Runner.Run([]string{"docker", "--version"})
	return result.Success
}

// Uninstall removes Docker from Ubuntu/Debian.
func (u *ubuntuInstaller) Uninstall(opts *installers.UninstallOptions) *installers.UninstallResult {
	u.CLI.PrintSection("Desinstallation de Docker")

	// Stop Docker service
	u.CLI.PrintInfo("Arret du service Docker...")
	u.Runner.Run([]string{"systemctl", "stop", "docker"}, runner.WithSudo())
	u.Runner.Run([]string{"systemctl", "disable", "docker"}, runner.WithSudo())

	// Remove Docker packages
	u.CLI.PrintInfo("Suppression des paquets Docker...")
	packages := []string{
		"docker-ce",
		"docker-ce-cli",
		"containerd.io",
		"docker-buildx-plugin",
		"docker-compose-plugin",
	}
	u.Runner.Run(
		append([]string{"apt-get", "purge", "-y"}, packages...),
		runner.WithSudo(),
	)
	u.CLI.PrintSuccess("Paquets Docker supprimes")

	// Remove Docker data
	if opts.RemoveData {
		u.CLI.PrintInfo("Suppression des donnees Docker...")
		dataDirectories := []string{
			"/var/lib/docker",
			"/var/lib/containerd",
			"/etc/docker",
			filepath.Join(u.SystemInfo.HomeDir, ".docker"),
		}

		for _, dir := range dataDirectories {
			u.Runner.Run([]string{"rm", "-rf", dir}, runner.WithSudo())
		}
		u.CLI.PrintSuccess("Donnees Docker supprimees")
	}

	// Remove repository
	u.Runner.Run([]string{"rm", "-f", "/etc/apt/sources.list.d/docker.list"}, runner.WithSudo())
	u.Runner.Run([]string{"rm", "-f", "/etc/apt/keyrings/docker.asc"}, runner.WithSudo())

	return installers.NewUninstallSuccessResult("Docker desinstalle avec succes")
}

File: docker_installer/installers/ubuntu.py
"""Ubuntu/Debian-specific Docker installer using official Docker repository."""

import os
from typing import Optional
from .base import BaseInstaller, InstallOptions, InstallResult
from ..utils.system import SystemInfo, OSType, Architecture
from ..utils.cli import CLI
from ..utils.runner import CommandRunner


class UbuntuInstaller(BaseInstaller):
    """Installer for Docker on Ubuntu/Debian using official Docker repository."""

    # Docker GPG key URL
    DOCKER_GPG_URL = "https://download.docker.com/linux/ubuntu/gpg"
    DOCKER_GPG_URL_DEBIAN = "https://download.docker.com/linux/debian/gpg"

    def __init__(self, system_info: SystemInfo, cli: CLI, runner: CommandRunner):
        """Initialize the Ubuntu Docker installer."""
        super().__init__(system_info, cli, runner)

    def _get_distro_codename(self) -> str:
        """Get the distribution codename for Docker repository."""
        result = self.runner.run(
            ["lsb_release", "-cs"],
            sudo=False
        )
        if result.success:
            return result.stdout.strip()

        # Fallback: try to read from /etc/os-release
        try:
            with open("/etc/os-release", "r") as f:
                for line in f:
                    if line.startswith("VERSION_CODENAME="):
                        return line.split("=")[1].strip().strip('"')
        except FileNotFoundError:
            pass

        # Default fallback
        return "jammy"  # Ubuntu 22.04

    def _get_arch_string(self) -> str:
        """Get the architecture string for Docker repository."""
        if self.system_info.architecture == Architecture.ARM64:
            return "arm64"
        return "amd64"

    def _remove_old_packages(self) -> bool:
        """Remove old Docker packages that might conflict."""
        old_packages = [
            "docker.io",
            "docker-doc",
            "docker-compose",
            "docker-compose-v2",
            "podman-docker",
            "containerd",
            "runc"
        ]

        self.cli.print_info("Suppression des anciens paquets Docker...")

        for package in old_packages:
            # Check if package is installed
            result = self.runner.run(
                ["dpkg", "-s", package],
                sudo=False
            )
            if result.success:
                self.runner.run(
                    ["apt-get", "remove", "-y", package],
                    sudo=True
                )

        return True

    def _install_prerequisites(self) -> bool:
        """Install prerequisites for Docker installation."""
        self.cli.print_info("Installation des prerequis...")

        packages = [
            "ca-certificates",
            "curl",
            "gnupg",
            "lsb-release"
        ]

        # Update apt
        result = self.runner.run(
            ["apt-get", "update"],
            description="Mise a jour des paquets",
            sudo=True,
            timeout=300
        )

        if not result.success:
            self.cli.print_warning("Impossible de mettre a jour apt")

        # Install packages
        result = self.runner.run(
            ["apt-get", "install", "-y"] + packages,
            description="Installation des prerequis",
            sudo=True,
            timeout=300
        )

        return result.success

    def _setup_docker_repository(self) -> bool:
        """Set up the official Docker repository."""
        self.cli.print_info("Configuration du repository Docker...")

        # Determine if Ubuntu or Debian
        is_debian = self.system_info.os_type == OSType.DEBIAN
        distro = "debian" if is_debian else "ubuntu"
        gpg_url = self.DOCKER_GPG_URL_DEBIAN if is_debian else self.DOCKER_GPG_URL

        # Create keyrings directory
        keyrings_dir = "/etc/apt/keyrings"
        self.runner.run(
            ["mkdir", "-p", keyrings_dir],
            sudo=True
        )

        # Download and add Docker GPG key
        gpg_file = f"{keyrings_dir}/docker.gpg"

        # Remove old key if exists
        if os.path.exists(gpg_file):
            self.runner.run(["rm", "-f", gpg_file], sudo=True)

        # Download GPG key
        result = self.runner.run(
            ["curl", "-fsSL", gpg_url, "-o", "/tmp/docker.gpg.tmp"],
            description="Telechargement de la cle GPG Docker",
            sudo=False
        )

        if not result.success:
            self.cli.print_error("Impossible de telecharger la cle GPG Docker")
            return False

        # Dearmor and install GPG key
        result = self.runner.run(
            ["gpg", "--dearmor", "-o", gpg_file, "/tmp/docker.gpg.tmp"],
            sudo=True
        )

        if not result.success:
            # Try alternative method
            result = self.runner.run(
                ["bash", "-c", f"cat /tmp/docker.gpg.tmp | gpg --dearmor -o {gpg_file}"],
                sudo=True
            )

        # Set proper permissions
        self.runner.run(
            ["chmod", "a+r", gpg_file],
            sudo=True
        )

        # Get architecture and codename
        arch = self._get_arch_string()
        codename = self._get_distro_codename()

        # Add Docker repository
        repo_line = f"deb [arch={arch} signed-by={gpg_file}] https://download.docker.com/linux/{distro} {codename} stable"

        result = self.runner.run(
            ["bash", "-c", f'echo "{repo_line}" > /etc/apt/sources.list.d/docker.list'],
            sudo=True
        )

        if not result.success:
            self.cli.print_error("Impossible d'ajouter le repository Docker")
            return False

        # Update apt with new repository
        result = self.runner.run(
            ["apt-get", "update"],
            description="Mise a jour avec le nouveau repository",
            sudo=True,
            timeout=300
        )

        return result.success

    def check_existing_installation(self) -> bool:
        """Check if Docker is already installed."""
        if self.runner.check_command_exists("docker"):
            # Verify docker daemon is accessible
            result = self.runner.run(
                ["docker", "info"],
                sudo=True,
                timeout=10
            )
            return result.success

        return False

    def install_docker(self) -> bool:
        """Install Docker Engine from official repository."""
        # Remove old packages
        self._remove_old_packages()

        # Install prerequisites
        if not self._install_prerequisites():
            self.cli.print_error("Impossible d'installer les prerequis")
            return False

        # Setup Docker repository
        if not self._setup_docker_repository():
            self.cli.print_error("Impossible de configurer le repository Docker")
            return False

        # Install Docker packages
        self.cli.print_info("Installation de Docker Engine...")

        docker_packages = [
            "docker-ce",
            "docker-ce-cli",
            "containerd.io",
            "docker-buildx-plugin",
            "docker-compose-plugin"
        ]

        result = self.runner.run(
            ["apt-get", "install", "-y"] + docker_packages,
            description="Installation de Docker",
            sudo=True,
            timeout=600
        )

        if not result.success:
            self.cli.print_error("Echec de l'installation de Docker")
            return False

        self.cli.print_success("Docker Engine installe")
        return True

    def install_docker_compose(self) -> bool:
        """Docker Compose v2 is installed as a plugin with Docker Engine."""
        # Check if already installed
        result = self.runner.run(
            ["docker", "compose", "version"],
            sudo=False
        )

        if result.success:
            self.cli.print_success("Docker Compose est deja installe")
            return True

        # If not installed, try to install the plugin separately
        result = self.runner.run(
            ["apt-get", "install", "-y", "docker-compose-plugin"],
            description="Installation de Docker Compose",
            sudo=True,
            timeout=300
        )

        return result.success

    def configure_docker(self) -> bool:
        """Configure Docker for the current user."""
        success = True

        if self.options and self.options.add_user_to_docker_group:
            # Add current user to docker group
            username = os.environ.get("USER") or os.environ.get("LOGNAME")

            if username and username != "root":
                self.cli.print_info(f"Ajout de l'utilisateur '{username}' au groupe docker...")

                # Check if docker group exists
                result = self.runner.run(
                    ["getent", "group", "docker"],
                    sudo=False
                )

                if not result.success:
                    # Create docker group
                    self.runner.run(
                        ["groupadd", "docker"],
                        sudo=True
                    )

                # Add user to group
                result = self.runner.run(
                    ["usermod", "-aG", "docker", username],
                    sudo=True
                )

                if result.success:
                    self.cli.print_success(f"Utilisateur '{username}' ajoute au groupe docker")
                    self.cli.print_warning("Deconnectez-vous et reconnectez-vous pour appliquer les changements")
                else:
                    self.cli.print_warning("Impossible d'ajouter l'utilisateur au groupe docker")
                    success = False

        if self.options and self.options.start_on_boot:
            # Enable Docker to start on boot
            self.cli.print_info("Configuration du demarrage automatique...")

            result = self.runner.run(
                ["systemctl", "enable", "docker.service"],
                sudo=True
            )

            if result.success:
                self.cli.print_success("Docker demarre automatiquement au boot")
            else:
                self.cli.print_warning("Impossible de configurer le demarrage automatique")

            # Also enable containerd
            self.runner.run(
                ["systemctl", "enable", "containerd.service"],
                sudo=True
            )

        return success

    def start_docker(self) -> bool:
        """Start Docker service."""
        # Check if already running
        result = self.runner.run(
            ["systemctl", "is-active", "docker"],
            sudo=False
        )

        if result.success and "active" in result.stdout:
            self.cli.print_success("Docker est deja en cours d'execution")
            return True

        # Start Docker service
        self.cli.print_info("Demarrage du service Docker...")

        result = self.runner.run(
            ["systemctl", "start", "docker"],
            sudo=True
        )

        if result.success:
            self.cli.print_success("Docker demarre")
            return True
        else:
            self.cli.print_error("Impossible de demarrer Docker")
            return False

    def install(self, options: InstallOptions) -> InstallResult:
        """Run the complete Docker installation for Ubuntu/Debian."""
        result = super().install(options)

        if result.success and options.add_user_to_docker_group:
            # Remind user about group changes
            result.warnings.append(
                "Deconnectez-vous et reconnectez-vous pour utiliser Docker sans sudo"
            )

        return result
